src/swish/io.ss

```diff
@@ -265,53 +269,126 @@
           (set-port-position! ip start)))
     (port-position ip))
 
-  (define osi-port-guardian (make-guardian))
-  (define osi-port-table (make-weak-eq-hashtable))
+  (define (sorted-cells ht create-time cell->record cell->handle)
+    ;; cell is in ht iff its cell->handle is not #f
+    (define (cell<? x1 x2)
+      (let ([t1 (create-time (cell->record x1))] [t2 (create-time (cell->record x2))])
+        (cond
+         [(< t1 t2) #t]
+         [(> t1 t2) #f]
+         [else (< (cell->handle x1) (cell->handle x2))])))
+    (let ([v (with-interrupts-disabled (hashtable-cells ht))])
+      (vector-sort! cell<? v)
+      v))
+
+  (define-record-type %type-reporter
+    (nongenerative)
+    (fields
+     (immutable name)
+     (immutable count)
+     (immutable print)))
+
+  (define type-reporters (make-hashtable symbol-hash eq?))
+
+  ;; TODO document expectation that handle is numeric
+  (define (add-type-reporter name table cell->record cell->handle get-create-time print-one)
+    (define (get-count) (hashtable-size table))
+    (define print
+      (case-lambda
+       [() (print (current-output-port))]
+       [(op)
+        (vector-for-each
+         (lambda (x)
+           (print-one op (cell->record x) (cell->handle x)))
+         (sorted-cells table get-create-time cell->record cell->handle))]))
+    (hashtable-update! type-reporters name
+      (lambda (prev)
+        (when prev
+          (errorf 'add-type-reporter "reporter for type ~s already exists" name))
+        (make-%type-reporter name get-count print))
+      #f))
+
+  (define (type-guardian-counts make-it set-it!)
+    (let ([it (make-it)])
+      (vector-for-each
+       (lambda (cell)
+         (set-it! it (car cell) ((%type-reporter-count (cdr cell)))))
+       (hashtable-cells type-reporters))
+      it))
+
+  (define (type-guardian-count type)
+    (match (hashtable-ref type-reporters type #f)
```

Comment:
Suggestion: match-let* is more efficient when there is only one clause to match.

Reply:
The same goes for type-guardian-print.

Reply:
Good idea. I modified the single-clause case of `match` to expand as `match-let*` since I often forget this and sometimes prefer the indentation of `match`. Hopefully I haven't missed anything in the transformation.